actors/builtin/reward/testing.go

```diff
@@ -0,0 +1,32 @@
+package reward
+
+import (
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/go-state-types/big"
+	"github.com/filecoin-project/specs-actors/v2/actors/builtin"
+	"github.com/filecoin-project/specs-actors/v2/actors/util/math"
+	"github.com/filecoin-project/specs-actors/v2/actors/util/adt"
+)
+
+type StateSummary struct{}
+
+var FIL = big.NewInt(1e18)
+var StorageMiningAllocationCheck = big.Mul(big.NewInt(1_100_000_000), FIL)
+
+func CheckStateInvariants(st *State, store adt.Store, priorEpoch abi.ChainEpoch, balance abi.TokenAmount) (*StateSummary, *builtin.MessageAccumulator, error) {
+	acc := &builtin.MessageAccumulator{}
+
+	acc.Require(big.Add(st.TotalStoragePowerReward, balance).Equals(StorageMiningAllocationCheck), "reward given + reward left != storage mining allocation")
+
+	acc.Require(st.Epoch == priorEpoch, "reward state epoch %d does not match priorEpoch %d", st.Epoch, priorEpoch)
+	acc.Require(st.EffectiveNetworkTime <= st.Epoch, "effective network time greater than state epoch")
+
+	acc.Require(st.CumsumRealized.LessThanEqual(st.CumsumBaseline), "cumsum realized > cumsum baseline")
+	acc.Require(st.CumsumRealized.GreaterThanEqual(big.Zero()), "cumsum realized < 0")
+	acc.Require(st.EffectiveBaselinePower.LessThanEqual(st.ThisEpochBaselinePower), "effective baseline power > baseline power")
+
+	computedBaseline := math.ExpBySquaring(BaselineExponent, int64(st.Epoch))
+	acc.Require(st.ThisEpochBaselinePower.Equals(computedBaseline), "state baseline power does not match computed")
```

Comment:
I expect this doesn't hold now, but do you expect it to after the upgrade? 

It probably won't hold immediately after an empty tipset.

Reply:
There is some predictable error between the two methods of computing the exponential so you're right it won't ever hold with equality.  I also wouldn't expect this on v0 state, only v2.  However our error is worse than the small predictable error because I was off by one while recomputing this value so I want to try `computedBaseline := math.ExpBySquaring(BaselineExponent, int64(priorEpoch))`.  Empty epochs should not affect this since we do one cron per epoch and always run `updateToNextEpoch` which keeps `ThisEpochBaselinePower` current.